source/extensions/filters/http/cors/cors_filter.cc

```diff
@@ -13,6 +13,21 @@ namespace Extensions {
 namespace HttpFilters {
 namespace Cors {
 
+Http::RegisterCustomInlineHeader<Http::RequestHeaderMap>
+    access_control_request_method(Http::Headers::get().AccessControlRequestMethod);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_allow_origin(Http::Headers::get().AccessControlAllowOrigin);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_allow_credentials(Http::Headers::get().AccessControlAllowCredentials);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_allow_methods(Http::Headers::get().AccessControlAllowMethods);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_allow_headers(Http::Headers::get().AccessControlAllowHeaders);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_max_age(Http::Headers::get().AccessControlMaxAge);
+Http::RegisterCustomInlineHeader<Http::ResponseHeaderMap>
+    access_control_expose_headers(Http::Headers::get().AccessControlExposeHeaders);
```

Comment:
This is what it looks like to an extension. Pretty nice I think. :)

Reply:
agreed this is nice, but it looks like this could be a static-init fiasco. Could this be done in a construct-on-first-use wrapper for each filter?

Or do we need the semantics to be that any filter that is linked in gets its custom headers registered before any packets flow?

Reply:
> Or do we need the semantics to be that any filter that is linked in gets its custom headers registered before any packets flow?

Yes, it has to work this way, similar to how extensions are registered. I don't think in practice there is any fiasco as all data is itself accessed using "fiasco proof" paradigms.

With that said, I *think* it might be possible to do something like allow extensions to register headers when they are statically registered somehow. I'm not sure if that is much better though. Either way I think this is something that could be potentially improved later if we go this direction, and later add the ability to register custom headers using CLI/bootstrap which I'm pretty sure is possible.